### ISSUE TITLE
Should Unchecked Casts be allowed in the bounds of Unchecked Functions

### ISSUE THREAD

**Issue by lenary** [Microsoft/checkedc-clang#311]
This is a semantics issue. The following currently doesn't work, because of the cast in the bounds declaration.

```
#pragma BOUNDS_CHECKED ON

_Unchecked
 void *memchr(const void *s : byte_count(n), int c, size_t n) :
  bounds(s, (char *) s + n);

#pragma BOUNDS_CHECKED OFF
```

The error is "error: invalid casting to unchecked pointer type in a checked scope or function".

Should it be allowed? My feeling would be "yes", but I worry that could break soundness. No need for an answer immediately, I'm just trying to get BOUNDS_CHECKED everywhere for our benchmarks

**Comment by dtarditi** [comment]
They don't need to be allowed.  You can just cast to an array_ptr<char> instead.

From a viewpoint of semantics, I don't think unchecked casts in bounds should be allowed in checked contexts.  Unchecked pointer arithmetic can result in undefined behavior according to C semantics.    It is undefined, for example, if a pointer points outsides a valid object (except for pointing one element past).   We may want to go one step further and ban unchecked pointer arithmetic from bounds expressions, even in unchecked contexts.

**Comment by lenary** [comment]
Ok, great, I'll go back and alter the checked headers later today. Expect a PR.

**Comment by lenary** [comment]
Closed by Microsoft/checkedc#165
